### PR TITLE
Otbn indirect reg

### LINKED ISSUE #4808: [otbn] Should we error out when incrementing "past 31" for WDR in BN.LID / BN.SID
When discussing the behaviour of increments, Philipp [suggested](https://github.com/lowRISC/opentitan/issues/3587#issuecomment-756871416) that we could raise an error if the WDR index was incremented past 31, since it probably shows a bug in the code.

The current spec allows code like this:
```
    addi x2, x0, 16
    addi x3, x0, 24
    addi x4, x0, 0x100
    loopi 8, 3
      bn.lid x2++, 0(x4)
      bn.lid x3++, 32(x4)
      addi x4, x4, 64
```
to load up "striped" data without too much pointer manipulation. The last iteration of that loop will wrap `x3` around from `0x1f` to zero. This wouldn't be allowed with the proposed change.

Should we raise an error in this case?

> estimate 8
> remaining 2021-03-08 4
> remaining 2021-03-22 1

Also need to change the RIG to not produce LID/SID code which uses register values larger than 31.

### PR DESCRIPTION
**Pull request by GregAC**
Fixes #4808

### CODE REVIEW

File: hw/ip/otbn/dv/otbnsim/sim/insn.py [hunk 1]
```diff
@@ -896,6 +902,11 @@ def execute(self, state: OTBNState) -> None:
         addr = (grs1_val + self.offset) & ((1 << 32) - 1)
 
         grs2_val = state.gprs.get_reg(self.grs2).read_unsigned()
+
+        if grs2_val > 0x1f:
+            state.stop_at_end_of_cycle(ILLEGAL_INSN)
+            return
```
Comment (GregAC) [inline hw/ip/otbn/dv/otbnsim/sim/insn.py:908]: This code works simulation wise (from some brief testing) but may not be great from a documentation psuedo-code perspective. Perhaps we'd prefer

```python
grs1_val = state.gprs.get_reg(self.grs1).read_unsigned()
addr = (grs1_val + self.offset) & ((1 << 32) - 1)

grs2_val = state.gprs.get_reg(self.grs2).read_unsigned()

if grs2_val <= 0x1f:
    wrs = grs2_val & 0x1f
    wrs_val = state.wdrs.get_reg(wrs).read_unsigned()
    
    state.dmem.store_u256(addr, wrs_val)
    
    if self.grs1_inc:
        new_grs1_val = (grs1_val + 32) & ((1 << 32) - 1)
        state.gprs.get_reg(self.grs1).write_unsigned(new_grs1_val)
    
    if self.grs2_inc:
        new_grs2_val = grs2_val + 1
        state.gprs.get_reg(self.grs2).write_unsigned(new_grs2_val)
else:
    raise_error(ILLEGAL_INSN)
```

**Comment by rswarbrick** [comment]
I'd rather not commit the second patch without someone doing a bit of work on the RIG (since it doesn't currently know to avoid having bits set in the top of the register). Is the first patch an unrelated cleanup? If so, maybe we should split that out and merge it immediately?

**Review by imphil — APPROVED** [review]
Thanks @GregAC! The behavior we have arrived there makes a nice ISA and a efficient implementation. I have some small comments on the implementation.

As Rupert said, let's get the RIG changes in as well if possible to avoid breaking our nightly regressions.

File: hw/ip/otbn/dv/otbnsim/sim/insn.py (hunk 1 above)
Comment (GregAC) [reply hw/ip/otbn/dv/otbnsim/sim/insn.py:908]: I'm happy to take a look, can do that as a follow-up PR so we can discuss appropriate substitutions there unless we're particularly keen to resolve this in this PR.

**Comment by GregAC** [comment]
First commit is spun out as https://github.com/lowRISC/opentitan/pull/5517

**Pull request merged**